### ISSUE TITLE
Tabular data format

### ISSUE THREAD

**Issue by RickMoynihan** [OpenGovIntelligence/json-qb#4]
If we're to support something like a `/table` route we'll need a data format to represent a table.

Broadly speaking there are two main approaches.  The essential difference between these styles is captured in the examples below (one array vs multiple arrays):
### 1. Use an Array for each column, and store each column in a map.

``` json
{
 "aberdeen":[1,2,3],
 "arbroath": [4,5,6]
}
```
### 2. Use a single Array of hashmaps, where each k/v pair is a column.

``` json
{
  "rows" : [{"aberdeen": 1, "arbroath": 4}, {"aberdeen": 2, "arbroath": 5} {"aberdeen": 3, "arbroath": 6}]
}
```

In practice both approaches benefit from additional metadata to represent things like sort order of columns, perhaps pagination etc...  So they would more likely become something like this:
### 1

``` json
{
  "cols": ["aberdeen","arbroath"],

  "data": {
   "aberdeen":[1,2,3],
   "arbroath": [4,5,6]
  }
  ,,,,
}
```
### 2

``` json
{
  "cols": ["aberdeen", "arbroath"],
  "rows" : [{"aberdeen": 1, "arbroath": 4}, {"aberdeen": 2, "arbroath": 5} {"aberdeen": 3, "arbroath": 6}]
 ,,,,
}
```

Approach 1 has the benefit of being more easily merged/combined with other data.  Adding/removing columns in particular is easier.  It is however harder to support very large datasets (at least without pagination).

Approach 2 is easier for the server to stream large quantities of data without consuming memory.  But is slightly harder to merge with additional columns client side, it's also harder to select a column (you essentially need to `map` a `getCell` function over it.

**Comment by BillSwirrl** [comment]
thanks @RickMoynihan - I realise these examples are not meant to be comprehensive, but just to note that they are missing an important bit of extra information, namely the value of whatever other dimension is free, over the three values provided for each area.

i.e. in "aberdeen":[1,2,3], that might correspond to eg year = 2001,2002,2003.  That information needs to go somewhere,  In the Google Datatable approach that could be attached to the columns, with a row for each area perhaps

**Comment by RickMoynihan** [comment]
Agreed @BillSwirrl.  I haven't yet proposed a complete example, the above two styles focus on what I consider the main difference between approaches.  

In practice both will need to be padded out with extra metadata, but I chose to omit that at this stage so we can focus on the essential differences.  When we decide on either 1 or 2, we can then start adapting them to include the extra info we need.  

Another topic is pagination, which we could probably lift from [JSON-api](http://jsonapi.org/format/#fetching-pagination), there's also other stuff in JSON-api which we might want to lift; perhaps wholesale.

**Comment by BillSwirrl** [comment]
Thanks Rick.  Agree on the idea of minimal illustrations of the overall style, but the values of the time dimension (in my example) are a fundamental thing we need, so I think it would be worth having data structure examples that include something like that, to help us to decide.  Other stuff like units or whatever I'm happy to class as 'extra metadata'.  I'll have a go at adapting those along those lines, just to help myself understand the details

**Comment by RickMoynihan** [comment]
@BillSwirrl Sorry, perhaps things are confused because I just copied guys example from slack; rather than writing them to model your example.  I still think what you're talking about is an orthogonal concern - one with many possible modelling approaches.  (not to propose one - but to illustrate one approach to mapping dimensions to cols/rows) in both `1` and `2` above:

`1.` 

``` json
{
  "cols": ["area", "2001", "2002", "2003"],
  "data": {
   "area": ["aberdeen", "arbroath", "brechin"],
   "2001" : [1,2,3],
   "2002" : [4,5,6],
   "2003":  [7,8,9]
  }
  ,,,,
}
```

`2.`

``` json
{
  "cols": ["area", "2001", "2002", "2003"],
  "rows" : [{"area": "aberdeen", "2001": 1, "2002": 4, "2003": 7}
               ,,,,]
 ,,,,
}
```

Both approaches above are still incomplete; as they'd need a way to state certain keys were the dimension data or measure data - but that's easy to do - I just didn't have the patience for handwriting the json.

Basically the essential difference between styles 1 and 2 that I'm perhaps badly trying to communicate isn't maps vs primitives for representing cells, it's multiple arrays vs one array.  For the record - I think I have a preference for approach 1 here, as we're going to add pagination anyway.

**Comment by zeginis** [comment]
There is also the approach that json-stat uses. It uses one array for the data but defines the dimension values only at the description of the table, not inside the data. The data are interpreded in a [row-major order](https://en.wikipedia.org/wiki/Row-major_order)

```
 {
  "cols": [ "2001", "2002", "2003"],
  "rows": ["aberdeen", "arbroath", "brechin"]
  "data": {[1,2,3,4,5,6,7,8,9]
 }
```

If we follow the json-stat format then this could be writen as:

```
{
"id" : ["geo","time"],
"size" : [3,3],
"dimension":{
      "geo" : {
         "category" : {"index" : ["aberdeen", "arbroath", "brechin"]}
       },
       "time" : {
         "category" : {"index" : ["2001", "2002", "2003"]}
       }
  },
"value" : [ 1,2,3,4,5,6,7,8,9]
 }
```

**Comment by RickMoynihan** [comment]
Closing this issue as we have something more concrete to speak about, which we can raise specific issues on.
